config/internal/configsource/manager.go

```diff
@@ -305,6 +313,109 @@ func (m *Manager) expandConfigSources(ctx context.Context, s string) (interface{
 	return retrieved.Value(), nil
 }
 
+// expandString expands environment variables and config sources that are specified on the string.
+func (m *Manager) expandString(ctx context.Context, s string) (interface{}, error) {
+	// Code based on os.Expand function. All delimiters that are checked against are
+	// ASCII so bytes are fine for this operation.
+	var buf []byte
+
+	// Using i, j, and w variables to keep correspondence with os.Expand code.
+	// i tracks the index in s from which a slice to be appended to buf should start.
+	// j tracks the char being currently checked and also the end of the slice to be appended to buf.
+	// w tracks the number of characters being consumed after a prefix identifying env vars or config sources.
+	i := 0
+	for j := 0; j < len(s); j++ {
+		if s[j] == expandPrefixChar && j+1 < len(s) {
+			if buf == nil {
+				buf = make([]byte, 0, 2*len(s))
+			}
+			buf = append(buf, s[i:j]...)
+
+			var retrieved interface{}
+			var err error
+			w := 0 // number of bytes consumed on this pass
+
+			switch {
+			case s[j+1] == expandPrefixChar:
+				// Escaped prefix:
+				buf = append(buf, s[j])
+				w = 1 // consumed a single char
+
+			case s[j+1] == '{':
+				// Bracketed usage, consume everything until first '}' exactly as os.Expand.
+				var content string
+				content, w = getShellName(s[j+1:])
+				content = strings.Trim(content, " ") // Allow for some spaces.
+				if len(content) > 1 && content[0] == expandPrefixChar {
+					name, _ := getShellName(content[1:])
+					if cfgSrc, ok := m.configSources[name]; ok {
```

Comment:
So we are breaking the old config that is using environment names which coincide with config source names, right? This will potentially also break more config over time as new config sources are added. I am not sure we want to do this.
Can we avoid trying to interpret this as a config source and then fallback to env variable name? Can we just interpret this as an env var name as it was in the past and never treat it as a config source name?

Reply:
> So we are breaking the old config that is using environment names which coincide with config source names, right?

Correct. However, I tend to think that the same name issue is an unlikely problem, keeping in mind that since the user can name the config source there are ways to avoid it without changing env vars.

> Can we avoid trying to interpret this as a config source and then fallback to env variable name?

To do that we probably have to limit the usage of config sources to be in the form of bracketed single-line and multi-line. Let me play around with it.

Reply:
Peeking after the name for ':' and treating it as config source only in that case.